Upgrade to ffmpeg v4.4 and use new detector pre-loading API

jailuthra opened the pull request:
**What does this pull request do? Explain your changes. (required)**
<!-- A clear and concise description of what this pull request does. -->

Upgrade ffmpeg and LPMS to use v4.4, a better detector API and misc. transcoding fixes.

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->

https://github.com/livepeer/go-livepeer/pull/1979/commits/81a4e06be0c2f7ec30702b5753c205a0bbc082c7
In this commit we switch to the new session pre-loading API introduced in https://github.com/livepeer/lpms/pull/246#issuecomment-891131495

Earlier we used to initialize FFmpeg at the time of node startup using `InitFFmpegWithDetectorProfile` - which would internally load a tensorflow model and share it across instances of our scene classification filter. We did that as it was an expensive (~3s) operation that couldn't be done when a new stream came in.

The earlier approach was problematic because we shared (a non-thread-safe) Tensorflow graph across concurrent go-routines, and it was inflexible as we could only initialize one detection model per node.

In this commit, I've mostly sticked close to the legacy idea of "pre-loading a single model" at node startup, but using the newer transcode-session API. Whenever a new stream comes in, we now initialize a new detector session, which has an impact of (~500ms) on the transcode time of the first segment.

This meets the real-time threshold on my local testing, but the plan is to improve the load balancer to have a constant `MinPreinitializedDetectorSessions` always ready at hand before a stream comes in. Doing that will reduce the additional overhead introduced by the new API, while also preventing any thread-safety issues in the Tensorflow models.

*See the commit history for rest*

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->
- [x] Local benchmarks and B-O-T Testing
- [x] Test in staging
- [x] Test in a single production region

**Does this pull request close any open issues?**
<!-- Fixes # -->
Fixes #1962 

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Read the [contribution guide](./doc/contributing.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [x] README and other documentation updated
- [x] [Pending changelog](./CHANGELOG_PENDING.md) updated

jailuthra committed:
install_ffmpeg,mod: Update LPMS & ffmpeg to v4.4

jailuthra committed:
livepeer_bench: Use new detector session API

Also use ConstrainedHigh H264 profile in our config to disable B-frames,
keeping it same as how it was before upgrading ffmpeg to v4.4.

jailuthra committed:
core,cmd/livepeer: Use new detector session API

yondonfu commented on core/lb.go:
```diff
@@ -14,17 +14,21 @@ import (
 
 var ErrTranscoderBusy = errors.New("TranscoderBusy")
 var ErrTranscoderStopped = errors.New("TranscoderStopped")
+var DetectorProfile ffmpeg.DetectorProfile
```
Can we add a comment here as a reminder that this is just for convenience for now since we only support pre-loading the model for a single detector profile?

jailuthra replied:
Will do!

yondonfu commented on cmd/livepeer_bench/livepeer_bench.go:
```diff
@@ -104,17 +104,24 @@ func main() {
 	}
+
+	ffmpeg.InitFFmpegWithLogLevel(ffmpeg.FFLogWarning)
+	var detector_tc *ffmpeg.Transcoder
```
Nit: Camel case is more conventional.

yondonfu commented on cmd/livepeer_bench/livepeer_bench.go:
```diff
@@ -126,7 +133,23 @@ func main() {
+					end := time.Now()
+					fmt.Printf("InitDetectorSession time %0.4v for session %v\n", end.Sub(t).Seconds(), i)
+					if err != nil {
+						glog.Fatalf("Could not initialize detector profiles")
```
```suggestion
						glog.Fatalf("Could not initialize detector")
```

yondonfu commented on cmd/livepeer_bench/livepeer_bench.go:
```diff
@@ -126,7 +133,23 @@ func main() {
-			tc := ffmpeg.NewTranscoder()
+			var tc *ffmpeg.Transcoder
+			if *detectionFreq > 0 {
+				if detector_tc != nil {
```
Isn't this set earlier on L115?

jailuthra replied:
We set it to `nil` a couple lines below here - so that from the 2nd session onwards we create a new session on-the-fly.

Alternatively, I think we should match it to what we do on the livepeer node - i.e. ignore the pre-initialized session altogether and init a new one each time.

Will fix!

jailuthra commented on core/lb.go:
```diff
@@ -80,8 +89,18 @@ func (lb *LoadBalancingTranscoder) createSession(md *SegTranscodingMetadata) (*t
+	var lpmsSession TranscoderSession
+	if md.DetectorEnabled {
+		var err error
+		lpmsSession, err = lb.newDetectorT(DetectorProfile, transcoder)
```
Hardcode detection device to "0"
Thanks @yondonfu for catching this!

jailuthra replied:
Made this change in https://github.com/livepeer/go-livepeer/pull/1979/commits/c17d16e113775f0957bcaefb41a65b66dd756ea6 in `core/transcoder.go` as we'd still want to schedule the transcoding on the GPU that load balancer selected - but only the detection filter hardcoded to device "0".

jailuthra committed:
core/transcoder: Hardcode detection device to 0

yondonfu commented on core/transcoder.go:
```diff
@@ -145,6 +145,16 @@ func NewNvidiaTranscoder(gpu string) TranscoderSession {
+func NewNvidiaTranscoderWithDetector(detector ffmpeg.DetectorProfile, gpu string) (TranscoderSession, error) {
+	// Hardcode detection to device 0 for now
+	// Transcoding can still run on a separate GPU as we copy frames to CPU before detection
+	session, err := ffmpeg.NewTranscoderWithDetector(detector, "0")
```
Just a note that can be addressed outside of this PR:

I think this API is a bit confusing. While I know that the device ID arg is only relevant for initializing the DNN filtergraph, when I first read this function my first thought is that the device ID would be used for both the transcoder and the DNN filtergraph.

yondonfu commented on install_ffmpeg.sh:
```diff
@@ -137,8 +137,8 @@ if [ ! -e "$ROOT/ffmpeg/libavcodec/libavcodec.a" ]; then
     --enable-parser=aac,aac_latm,h264 \
-    --enable-filter=abuffer,buffer,abuffersink,buffersink,afifo,fifo,aformat \
-    --enable-filter=aresample,asetnsamples,fps,scale,lvpdnn \
+    --enable-filter=abuffer,buffer,abuffersink,buffersink,afifo,fifo,aformat,format \
```
I might've missed this in the LPMS PR review - what is the `format` filter used for?

jailuthra replied:
The `format` filter is needed after `hwdownload` filter to signal that we expect downloaded CUDA frames are in `NV12` pixel format 
https://ffmpeg.org/ffmpeg-filters.html#hwdownload-1 used [here in lpms](https://github.com/livepeer/lpms/commit/e85ff00f8df274bacc9bca2a154493800979b275#diff-4ab68a4de67768f1acb5be217d4ec78bab27b99381025beedf82c11a5fa9c7aeR310)

yondonfu replied:
Ah right! Thanks for the explanation.

yondonfu reviewed (approved):
Just a few non-blocking questions, but LGTM

yondonfu reviewed (commented):
Mind also adding an entry to PENDING_CHANGELOG?

jailuthra committed:
CHANGELOG_PENDING: Add entry for #1979

jailuthra commented:
Tested the last fix in a multi-GPU setup and it worked as expected. Rebased and added a PENDING_CHANGELOG entry.

Holding on a merge until the new go-livepeer release is officially out.

yondonfu commented:
> Tested the last fix in a multi-GPU setup and it worked as expected.

🥳 🎉 

> Holding on a merge until the new go-livepeer release is officially out.

The release was just published so good to merge after resolving the conflict 👍 .

yondonfu commented:
Btw I forgot to clear CHANGELOG_PENDING in https://github.com/livepeer/go-livepeer/pull/1986, but looks like the commit with the CHANGELOG_PENDING entry will clear all the old entries anyway 👍 .
